src/librustdoc/html/sources.rs

```diff
@@ -98,16 +99,20 @@ struct SourceCollector<'a, 'tcx> {
     emitted_local_sources: FxHashSet<PathBuf>,
 }
 
-impl DocFolder for SourceCollector<'_, '_> {
-    fn fold_item(&mut self, item: clean::Item) -> Option<clean::Item> {
+impl DocVisitor for SourceCollector<'_, '_> {
+    fn visit_item(&mut self, item: &clean::Item) {
+        if !self.cx.include_sources {
+            return;
```

Comment:
This line of code cries out for a more explicit way to do control flow than this. This conditional seems to stop it from trying to recurse into child nodes, but it can't stop the parent from trying to visit all of the siblings (and that seems wasteful), and I really only figured that out by carefully reading the implementation.

The API used by [`TypeVisitor`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/fold/trait.TypeVisitor.html) seems like a good source for plagiarizing, since `ControlFlow::Break` provides a way to "just bail out" like we want here.

Reply:
Yeah, `ControlFlow` makes everything better 🙃 

I'd be happy to try out switching `DocVisitor` to use `ControlFlow`, but that'd make this change even bigger, so I think let's save it for a future PR.